cpp/src/utilities/device_atomics.cuh

```diff
@@ -0,0 +1,517 @@
+                reinterpret_cast<T_int*>(addr),
+                type_reinterpret<T_int, T>(compare),
+                type_reinterpret<T_int, T>(update_value));
+
+            return type_reinterpret<T, T_int>(ret);
+        }
+    };
+
+    // call atomic function with type cast between same underlying type
+    template <typename T, typename Functor>
+    __forceinline__  __device__
+    T typesAtomicOperation32(T* addr, T val, Functor atomicFunc)
+    {
+        using T_int = int;
+        T_int ret = atomicFunc(reinterpret_cast<T_int*>(addr),
+            cudf::detail::type_reinterpret<T_int, T>(val));
+
+        return cudf::detail::type_reinterpret<T, T_int>(ret);
+    }
+
+    // call atomic function with type cast between same underlying type
+    template <typename T, typename Functor>
+    __forceinline__  __device__
+    T typesAtomicOperation64(T* addr, T val, Functor atomicFunc)
+    {
+        using T_int = long long int;
+        T_int ret = atomicFunc(reinterpret_cast<T_int*>(addr),
+            cudf::detail::type_reinterpret<T_int, T>(val));
+
+        return cudf::detail::type_reinterpret<T, T_int>(ret);
+    }
+
+} // namespace detail
+
+
+template <typename T, typename BinaryOp>
+__forceinline__  __device__
+T genericAtomicOperation(T* address, T const & update_value, BinaryOp op)
+{
+    return cudf::detail::genericAtomicOperationImpl<T, BinaryOp, sizeof(T)>()
+        (address, update_value, op);
+}
+
+// ------------------------------------------------------------------------
+// Binary ops for sum, min, max
+struct DeviceSum {
+    template<typename T>
+    __device__
+    T operator() (const T &lhs, const T &rhs) {
+        return lhs + rhs;
+    }
+
+    template<typename T>
+    static constexpr T identity() { return T{0}; }
+};
+
+struct DeviceMin{
+    template<typename T>
+    __device__
+    T operator() (const T &lhs, const T &rhs) {
+        return lhs <= rhs? lhs: rhs;
+    }
+
+    template<typename T>
+    static constexpr T identity() { return std::numeric_limits<T>::max(); }
+};
+
+struct DeviceMax{
+    template<typename T>
+    __device__
+    T operator() (const T &lhs, const T &rhs) {
+        return lhs >= rhs? lhs: rhs;
+    }
+
+    template<typename T>
+    static constexpr T identity() { return std::numeric_limits<T>::lowest(); }
+};
+
+} // namespace cudf
+
+// ------------------------------------------------------------------------
+// ------------------------------------------------------------------------
+// Overloads for `atomicAdd`.
+// Cuda supports `sint32`, `uint32`, `uint64`, `float`, `double`
+
+__forceinline__ __device__
+int8_t atomicAdd(int8_t* address, int8_t val)
+{
+    return cudf::genericAtomicOperation(address, val, cudf::DeviceSum{});
+}
+
+__forceinline__ __device__
+int16_t atomicAdd(int16_t* address, int16_t val)
+{
+    return cudf::genericAtomicOperation(address, val, cudf::DeviceSum{});
+}
+
+__forceinline__ __device__
+int64_t atomicAdd(int64_t* address, int64_t val)
```

Comment:
I think it would also be nice to add overloads for all 64-bit types including `double` on CUDA_ARCH<600, to enable compilation for Maxwell GPUs.

Reply:
See https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomic-functions

Reply:
Okay, `atomicADD` for `double` is supported from Pascal.
Add it at latest commit.

Reply:
Fantastic.  Nice work.